packages/entity-server/src/routes/hierarchy/EntityRelationsRoute.ts

```diff
@@ -0,0 +1,150 @@
+/**
+ * Tupaia
+ * Copyright (c) 2017 - 2021 Beyond Essential Systems Pty Ltd
+ */
+
+import { Route } from '@tupaia/server-boilerplate';
+import { reduceToDictionary, reduceToArrayDictionary } from '@tupaia/utils';
+import { EntityType } from '../../models';
+import {
+  HierarchyRequest,
+  HierarchyRequestParams,
+  HierarchyRequestBody,
+  HierarchyRequestQuery,
+  EntityResponseObject,
+} from './types';
+
+export type RelationsRequest = HierarchyRequest<
+  HierarchyRequestParams,
+  | (EntityResponseObject & { descendants: EntityResponseObject[] })[] // groupBy: ancestor, flattenToCode: false
+  | (EntityResponseObject & { ancestor: EntityResponseObject })[] // groupBy: descendant, flattenToCode: false
+  | Record<string, string[]> // groupBy: ancestor, flattenToCode: true
+  | Record<string, string>, // groupBy: descendant, flattenToCode: true
+  HierarchyRequestBody,
+  HierarchyRequestQuery & {
+    ancestorType?: string;
+    descendantType?: string;
+    groupBy?: 'ancestor' | 'descendant';
+    flattenToCode?: 'true' | 'false';
```

Comment:
Thoughts on the naming of these params `flattenToCode` and `groupBy`? Couldn't come up with anything better but they don't feel quite right to me either...

Reply:
We could rename `flattenToCode` to just `flat`, whose default behaviour is using `code` fields. This allows for other fields to be used when flatting in the future - with the addition of an extra param, eg `field`/`flatField`

We may never need to do that, but it's good to have the option!

Reply:
I agree that `groupBy` doesn't feel exactly right, although it's not too bad either. I'll read the code and try to think of other options!

Reply:
Couldn't think of anything better as of yet - but `groupBy` doesn't sound too bad!